pkg/types/tuf/v0.0.1/entry.go

```diff
@@ -0,0 +1,411 @@
+	if err != nil {
+		return nil, err
+	}
+	canonicalEntry.Root.Version = v.TufObj.Root.Version
+	canonicalEntry.Root.Expires = v.TufObj.Root.Expires
+	canonicalEntry.Root.Type = v.TufObj.Root.Type
+
+	canonicalEntry.Manifest = &models.TufManifestV001Schema{Signed: &models.TufManifestV001SchemaSigned{}}
+	canonicalEntry.Manifest.Signed.Content, err = v.sigObj.CanonicalValue()
+	if err != nil {
+		return nil, err
+	}
+	canonicalEntry.Manifest.Version = v.TufObj.Manifest.Version
+	canonicalEntry.Manifest.Expires = v.TufObj.Manifest.Expires
+	canonicalEntry.Manifest.Type = v.TufObj.Manifest.Type
+
+	// wrap in valid object with kind and apiVersion set
+	tuf := models.Tuf{}
+	tuf.APIVersion = swag.String(APIVERSION)
+	tuf.Spec = &canonicalEntry
+
+	bytes, err := json.Marshal(&tuf)
+	if err != nil {
+		return nil, err
+	}
+
+	return bytes, nil
+}
+
+// Validate performs cross-field validation for fields in object
+// FIXME: we can probably export ValidateMetablock on in-toto.go
+func (v V001Entry) Validate() error {
+	root := v.TufObj.Root
+	if root == nil || root.Signed == nil {
+		return errors.New("missing root")
+	}
+	if len(root.Signed.Content) == 0 && root.Signed.URL.String() == "" {
+		return errors.New("root must be specified")
+	}
+
+	tufManifest := v.TufObj.Manifest
+	if tufManifest == nil || tufManifest.Signed == nil {
+		return errors.New("missing TUF metadata")
+	}
+	if len(tufManifest.Signed.Content) == 0 && tufManifest.Signed.URL.String() == "" {
+		return errors.New("TUF metadata must be specified")
+	}
+	return nil
+}
+
+func (v *V001Entry) Attestation() (string, []byte) {
+	return "", nil
+}
+
+func createTufManifest(manifest []byte) (*models.TufManifestV001Schema, error) {
+	re := &models.TufManifestV001Schema{}
+	s := &data.Signed{}
+	if err := json.Unmarshal(manifest, s); err != nil {
+		return nil, err
+	}
+	baseSigned := &BaseSigned{}
+	if err := json.Unmarshal(s.Signed, baseSigned); err != nil {
+		return nil, err
+	}
+	// Canonicalize the JSON of the entire manifest
+	var decoded map[string]interface{}
+	if err := json.Unmarshal(manifest, &decoded); err != nil {
+		return nil, err
+	}
+	msg, err := cjson.Marshal(decoded)
+	if err != nil {
+		return nil, err
+	}
+	re.Signed = &models.TufManifestV001SchemaSigned{Content: strfmt.Base64(msg)}
+	re.Expires = baseSigned.Expires.String()
+	re.Type = baseSigned.Type
+	re.Version = int64(baseSigned.Version)
+	return re, nil
+}
+
+func (v V001Entry) CreateFromArtifactProperties(ctx context.Context, props types.ArtifactProperties) (models.ProposedEntry, error) {
+	// This will do only syntactic checks of the metablock, not signature verification.
+	// Signature verification occurs in FetchExternalEntries()
+	returnVal := models.Tuf{}
+	re := V001Entry{}
+	re.TufObj.Version = "1.0.0" // TODO: Get tuf specification for root manifest
+
+	// we will need the manifest and root
+	// Not sure why but only ArtifactPath is ever set with a filepath name
```

Comment:
re this comment: `props.ArtifactPath` is a `url.URL` so if `props.ArtifactPath.IsAbs()` is `true`, then it should be handled as a URL to be fetched. 

see https://github.com/sigstore/rekor/blob/557dde02c69090da0d9e506bd0a34bc49e5a42a5/pkg/types/rekord/v0.0.1/entry.go#L402 for an example

Reply:
AH! Also fixed it so that it actually handles the paths properly when ArtifactPath is set.